[WIP] Add first pass at db router [#OSF-7406]

cwisecarver opened the pull request:
<!-- Before submit your Pull Request, make sure you picked
     the right branch:

     - For hotfixes, select "master" as the target branch
     - For new features, select "develop" as the target branch
     - For release feature fixes, select the relevant release branch (release/X.Y.Z) as the target branch -->

## Purpose

<!-- Describe the purpose of your changes -->

## Changes

<!-- Briefly describe or list your changes  -->

## Side effects

<!--Any possible side effects? -->


## Ticket

<!-- Link to JIRA ticket, if applicable e.g. https://openscience.atlassian.net/browse/OSF-1234 -->

icereval commented on osf/db/router.py:
```diff
@@ -0,0 +1,70 @@
+        for name, dsn in self.DSNS.iteritems():
+            conn = self._get_conn(dsn)
+            cur = conn.cursor()
+            cur.execute('SELECT pg_is_in_recovery();')
```
This is good but lets stick with the JDBC implementation if at all possible.

https://github.com/pgjdbc/pgjdbc/blob/master/pgjdbc/src/main/java/org/postgresql/core/v3/ConnectionFactoryImpl.java#L692

cwisecarver replied:
They're probably the same. My only concern would be a pedantic one. Yours checks if the server is read only. Mine checks to see if one is master vs one is slave. You can technically have a master that's read only. It's going to error either way though if you try to write to it.

cwisecarver commented:
@icereval @chrisseto  Updated the query, added some comments, added setting to enable router, still running migrations so haven't actually tested running django with it.

mfraezz commented:
Closing in favor of #6820
